packages/cw22/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+/*!
+CW22 defines a way for a contract to declare which interfaces do the contract implement
+This standard is inspired by the EIP-165 from Ethereum. Originally it was proposed to
+be merged into CW2: Contract Info, then it is splitted to a separated cargo to keep CW2
+being backward compatible.
+
+Each supported interface contains a string value pointing to the corresponding cargo package
+and a specific release of the package. There is also a function to check whether the contract
+support a specific version of an interface or not.
+
+The version string for each interface follows Semantic Versioning standard. More info is in:
+https://docs.rs/semver/latest/semver/
+*/
+
+mod query;
+
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::StdError;
+use cosmwasm_std::{StdResult, Storage};
+use cw_storage_plus::Map;
+use query::VersionResponse;
+use semver::{Version, VersionReq};
+
+pub const SUPPORTED_INTERFACES: Map<String, String> = Map::new("supported_interfaces");
+
+#[cw_serde]
+pub struct ContractSupportedInterface {
+    /// supported_interface is an optional parameter returning a vector of string represents interfaces
+    /// that the contract support The string value is the interface crate names in Rust crate Registry.
+    /// This parameter is inspired by the EIP-165 from Ethereum.
+    /// Each string value should follow a common standard such as <Registry Domain>:<Crate Name>
+    /// e.g "crates.io:cw2"
+    /// NOTE: this is just a hint for the caller to adapt on how to interact with this contract.
+    /// There is no guarantee that the contract actually implement these interfaces.
+    pub supported_interface: String,
+    /// semantic version on release tags of the interface package following SemVer guideline.
+    /// e.g  "0.16.0"
+    pub version: String,
+}
+
+/// set_contract_supported_interface should be used in instantiate to store the original version
+/// of supported interfaces. It should also be used after every migration.
+pub fn set_contract_supported_interface(
+    store: &mut dyn Storage,
+    mut supported_interfaces: Vec<ContractSupportedInterface>,
+) -> Result<(), StdError> {
+    while let Some(supported_interface) = supported_interfaces.pop() {
+        let id = supported_interface.supported_interface;
```

Comment:
Why not just use a shorter local variable than "supported_interface" and inline this.  The following should be clear enough (but change x to item or so if you prefer)

```rust
for x in supported_interfaces {
  SUPPORTED_INTERFACES.save(store, &x.supported_interface, &x.version)?;
}
```

Reply:
Actually, since we assume all stored versions are semver later on in queries, enforce it here.
Parse the values as Version and error if they are invalid - much better to error on instantiate (fail fast in tests) than later when first querying in production.